### PR TITLE
boards: arm: gd32f407v_start: Add initial support for gd32f407v_start

### PR DESCRIPTION
**Pull request by soburi**
Add initial support for the GD32F407V-START board.

This patch is based on GD32E507V-START's one.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

### CODE REVIEW

**Comment by soburi** [comment]
I found using openocd for debugging GD32F407xx in the platform-io package (it seems community base made).
But, I don't include it because I can't verify that it is supported at this time.

https://github.com/CommunityGD32Cores/platform-gd32/blob/52cb1be23bcff41a2b9da6adf3ce9bbdd9b3bf15/boards/gd32407v_start.json#L26

**Comment by nandojve** [comment]
Hi @soburi ,

Update hash.

**Comment by soburi** [comment]
Hi @nandojve ,

> Hi @soburi ,
> 
> Update hash.

Rebase done. All CI passed. 👍

File: soc/arm/gigadevice/gd32f4xx/Kconfig.defconfig.gd32f407 [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) 2022 TOKITA Hiroshi <tokita.hiroshi@gmail.com>
```
Comment (cameled) [inline soc/arm/gigadevice/gd32f4xx/Kconfig.defconfig.gd32f407:1]: Please consider to split this commit to `dts:` and `soc:`.

File: dts/arm/gigadevice/gd32f4xx/gd32f407Xe.dtsi [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/*
```
Comment (cameled) [inline dts/arm/gigadevice/gd32f4xx/gd32f407Xe.dtsi:1]: I found you give a upper case for the general character `X`, but we already have a `gd32f470ik.dtsi` file. And for gd32, general character `x` is lower case.
So, please consider to change the filename.

File: boards/arm/gd32f407v_start/gd32f407v_start.dts [hunk 3]
```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2022, TOKITA Hiroshi <tokita.hiroshi@gmail.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/dts-v1/;
+
+#include <gigadevice/gd32f4xx/gd32f407Xk.dtsi>
```
Comment (cameled) [inline boards/arm/gd32f407v_start/gd32f407v_start.dts:8]: General character `X` to `x`.

**Review by cameled — COMMENTED** [review]
Just some format issue need to confirm.

File: soc/arm/gigadevice/gd32f4xx/Kconfig.defconfig.gd32f407 (hunk 1 above)
Comment (soburi) [reply soc/arm/gigadevice/gd32f4xx/Kconfig.defconfig.gd32f407:1]: Thank you for reviewing.
Fixed it.

File: dts/arm/gigadevice/gd32f4xx/gd32f407Xe.dtsi (hunk 2 above)
Comment (soburi) [reply dts/arm/gigadevice/gd32f4xx/gd32f407Xe.dtsi]: Correct filenames.

File: boards/arm/gd32f407v_start/gd32f407v_start.dts (hunk 3 above)
Comment (soburi) [reply boards/arm/gd32f407v_start/gd32f407v_start.dts]: Correct it same as filenames fixes.

File: dts/arm/gigadevice/gd32f4xx/gd32f407xe.dtsi [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+/ {
+	soc {
+		flash-controller@40023c00 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(512)>;
+			};
+		};
+	};
+};
```
Comment (gmarull) [inline dts/arm/gigadevice/gd32f4xx/gd32f407xe.dtsi:18]:
```suggestion
&flash0 {
	reg = <0x08000000 DT_SIZE_K(512)>;
};
```

File: dts/arm/gigadevice/gd32f4xx/gd32f407xg.dtsi [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+/ {
+	soc {
+		flash-controller@40023c00 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(1024)>;
+			};
+		};
+	};
+};
```
Comment (gmarull) [inline dts/arm/gigadevice/gd32f4xx/gd32f407xg.dtsi:18]:
```suggestion
&flash0 {
	reg = <0x08000000 DT_SIZE_K(1024)>;
};
```

File: dts/arm/gigadevice/gd32f4xx/gd32f407xk.dtsi [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+/ {
+	soc {
+		flash-controller@40023c00 {
+			flash0: flash@8000000 {
+				reg = <0x08000000 DT_SIZE_K(3072)>;
+			};
+		};
+	};
+};
```
Comment (gmarull) [inline dts/arm/gigadevice/gd32f4xx/gd32f407xk.dtsi:18]:
```suggestion
&flash0 {
	reg = <0x08000000 DT_SIZE_K(3072)>;
};
```

File: boards/arm/gd32f407v_start/board.cmake [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2022, TOKITA Hiroshi <tokita.hiroshi@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(jlink "--device=GD32F407VK" "--speed=4000")
+
+include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
```
Comment (gmarull) [inline boards/arm/gd32f407v_start/board.cmake:6]: this board should support openocd as well (just copy the support files from another f4xx board)

File: boards/arm/gd32f407v_start/doc/index.rst [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+follows:
+
+- JP3/4: Select 2-3 for both (boot from user memory)
+
+Using J-Link
+=============
+
+While the board comes with an embedded GD-Link programmer, it is not yet
+supported by OpenOCD. However, J-Link can be used to program the board using
+the SWD interface via PA13(SWDIO) and PA14(SWCLK) exposed in the JP6 header.
```
Comment (gmarull) [inline boards/arm/gd32f407v_start/doc/index.rst:87]: isn't it supported on f407 really?

File: dts/arm/gigadevice/gd32f4xx/gd32f407xe.dtsi (hunk 4 above)
Comment (soburi) [reply dts/arm/gigadevice/gd32f4xx/gd32f407xe.dtsi]: Fixed it.

File: dts/arm/gigadevice/gd32f4xx/gd32f407xg.dtsi (hunk 5 above)
Comment (soburi) [reply dts/arm/gigadevice/gd32f4xx/gd32f407xg.dtsi]: Fixed it.

File: dts/arm/gigadevice/gd32f4xx/gd32f407xk.dtsi (hunk 6 above)
Comment (soburi) [reply dts/arm/gigadevice/gd32f4xx/gd32f407xk.dtsi]: Fixed it.

File: boards/arm/gd32f407v_start/doc/index.rst (hunk 8 above)
Comment (soburi) [reply boards/arm/gd32f407v_start/doc/index.rst]: I understand it as "openocd not OFFICIALLY support GD32".
https://sourceforge.net/p/openocd/mailman/message/37687417/

I think using pyOCD (#49594) is preferable.
So, I reworked this PR referencing #49594.

File: boards/arm/gd32f407v_start/board.cmake (hunk 7 above)
Comment (soburi) [reply boards/arm/gd32f407v_start/board.cmake]: See https://github.com/zephyrproject-rtos/zephyr/pull/49085#discussion_r956536474

**Review by gmarull — APPROVED** [review]

**Review by nandojve — APPROVED** [review]

**Review by cameled — APPROVED** [review]

**Pull request merged**